cmd/kube-apiserver/app/options/options.go

```diff
@@ -104,10 +107,12 @@ func NewServerRunOptions() *ServerRunOptions {
 		Metrics:                 metrics.NewOptions(),
 		Logs:                    logs.NewOptions(),
 
-		EnableLogsHandler:      true,
-		EventTTL:               1 * time.Hour,
-		MasterCount:            1,
-		EndpointReconcilerType: string(reconcilers.LeaseEndpointReconcilerType),
+		EnableLogsHandler:         true,
+		EventTTL:                  1 * time.Hour,
+		MasterCount:               1,
+		EndpointReconcilerType:    string(reconcilers.LeaseEndpointReconcilerType),
+		LeaseDurationSeconds:      3600,
+		LeaseRenewIntervalSeconds: 10,
```

Comment:
I guess I expected the cleanup code to wait an hour past the expiration, rather than making the lease last an hour?

Reply:
The Lease API is misused here. It's more of a heartbeat object. 

I assume you're thinking:
- apiserver acquires a lease with a smaller duration (e.g. 1 minute), during the period no one (including the apiserver itself?) can force acquire the lease
- after the previous lease expires, the apiserver should try periodically re-acquiring the lease until success (10s)
- otherwise, the cleanup code will delete the expired lease after an hour.

Here the lease controller re-acquires the lease before the previous lease expires. It's been like this in the [node lease controller](https://github.com/kubernetes/kubernetes/blob/f75316e294c4c1ccc4afc0d89c78bbde43859c0d/pkg/kubelet/kubelet.go#L789): the controller refreshes the lease (every 10s) before the lease expires (40s).

Reply:
> I expected the cleanup code to wait an hour past the expiration

Since a lease gets acquired before it gets expired, the expiration means the process failed to send heartbeat within `LeaseDurationSeconds`. I think letting the cleanup code to wait for "the process failed to send heartbeat within `LeaseDurationSeconds` + 1 hour" introduces two parameters that do the same thing. 

That said, in the other logic (https://github.com/kubernetes/kubernetes/pull/95533#discussion_r517734159) the 1-hour wait would make sense. The cleanup criteria would be "the process failed to send heartbeat within 1 hour".

Reply:
Each apiserver is going to acquire a different lease, right? What cleans up an old lease when an apiserver reboots?

Reply:
Yes, each apiserver acquires a different lease. An apiserver-lease-garbage-collector (lives in kube-controller-manager) cleans up stale leases: https://github.com/kubernetes/kubernetes/pull/95895

Reply:
OK, made a comment there.